### PR TITLE
Add scan validator

### PR DESCRIPTION
**Pull request by zachfang**
- Add scan validator as the first step of filtering pipeline

### CODE REVIEW

File: src/scan_validator.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    const sensor_msgs::LaserScan& input_scan,
+    sensor_msgs::LaserScan &output_scan)
+{
+  // Copy input_scan data
+  output_scan = input_scan;
+
+  if(input_scan.ranges.size() != contour_.size())
+  {
+    ROS_WARN_THROTTLE(10.0, "Laserscan size doesn't match the validator contour, skip validation step");
+    return true;
```
Comment (zachfang) [inline src/scan_validator.cpp]: ROS_ERROR and return false and ROS_ERROR_ONCE

Comment (zachfang) [reply src/scan_validator.cpp]: Check other params for laserscan msg

File: src/scan_validator.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  int number_threshold = static_cast<int>(violation_percentage_ / 100.0 * input_scan.ranges.size());
```
Comment (zachfang) [inline src/scan_validator.cpp]: 100 rather 100.0

File: src/scan_validator.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  int cur_count = 0;
+
+  // Traverse each point
+  for(int i = 0; i < input_scan.ranges.size(); i++)
```
Comment (zachfang) [inline src/scan_validator.cpp]: auto or size_t

File: src/scan_validator.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  {
+    // avoid the case when the reading is 0.0
+    if(input_scan.ranges[i] < 0.01)
```
Comment (zachfang) [inline src/scan_validator.cpp]: 0.0 reading should count

File: src/scan_validator.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    {
+      cur_count += 1;
+    }
+  }
+
+  // If laser occlusion is caught, increase cur_scan_ or reset to zero otherwise
+  if(cur_count >= number_threshold) {
+    cur_scans_ += 1;
+    // Stop laserscan from propagating to next filter chain
+    if(cur_scans_ >= consecutive_scans_) {
```
Comment (zachfang) [inline src/scan_validator.cpp]: test one-scan trigger to see if there is false positive

File: src/scan_validator.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  const double EPS = 0.0001;
+
+  if(scan.ranges.size() != contour_.size() ||
+     !srs::BasicMath::equal<double>(scan.angle_min, angle_min_, EPS) ||
+     !srs::BasicMath::equal<double>(scan.angle_max, angle_max_, EPS) ||
+     !srs::BasicMath::equal<double>(scan.angle_increment, angle_increment_, EPS) ||
+     !srs::BasicMath::equal<double>(scan.range_min, range_min_, EPS) ||
+     !srs::BasicMath::equal<double>(scan.range_max, range_max_, EPS))
+  {
+    ROS_ERROR_THROTTLE(5.0, "Input laserscan configuration is different from defualt setting");
```
Comment (dantwinkler) [inline src/scan_validator.cpp]: Should be ROS_ERROR_ONCE

File: src/scan_validator.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  output_scan = input_scan;
+
+  // One-time check for laserscan configuration
+  if(!has_checked_laser_config_)
+  {
+    if(!checkLaserConfig(input_scan))
+    {
+      return false;
+    } else {
+      ROS_INFO("Input laserscan configuration matches defualt setting");
```
Comment (dantwinkler) [inline src/scan_validator.cpp]: If everything is ok, probably don't need to log an info here.  I think we assume things are ok, and only log on an error.

File: src/scan_validator.cpp [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+  return true;
+}
+
+bool laser_filters::ScanValidator::update(
+    const sensor_msgs::LaserScan& input_scan,
+    sensor_msgs::LaserScan &output_scan)
+{
+  // Copy input_scan data
+  output_scan = input_scan;
```
Comment (dantwinkler) [inline src/scan_validator.cpp]: Should probably only copy the data to the output when the input is valid (i.e. all true cases.).

**Review by dantwinkler — APPROVED** [review]

**Pull request merged**